### PR TITLE
Refactor burn-train

### PR DESCRIPTION
**Pull request by nathanielsimard**
Heavy refactor of `burn-train`.

## Changes

  * Added `LearnerComponents` to specify the abstraction boundaries necessary for the learner. It avoids having to specify most of the trait constraints in many places and simplifies the code.
  * Created `LearnerCheckpointer` to centralize checkpoint logic.
  * Renamed `LRScheduler` -> `LrScheduler` to follow the Rust standard.
  * Removed the `dashboard` within `burn-train`, which was a bit confusing. Replaced it with more appropriate names (`MetricsCallback`, `MetricsRenderer`, `MetricState`, `MetricUpdater`, `NumericMetricUpdater`, `TuiMetricsRenderer`, `CliMetricsRenderer`).
  * Removed `dyn` in Async Checkpointer.
  * Removed `dyn` and `mutex` in Async Callback.
  
## Breaking Changes

There are breaking changes, but most examples are untouched since type inference works the same way if used within the same function.

### CODE REVIEW

File: burn-train/src/learner/base.rs [hunk 1]
```diff
@@ -1,91 +1,67 @@
...
-    LR: LRScheduler,
-{
-    pub(super) fn checkpoint(
-        model: &Model,
-        optim: &Optim,
-        scheduler: &LR,
-        checkpointer_model: &CheckpointModel<Model, B>,
-        checkpointer_optimizer: &CheckpointOptim<Optim, Model, B>,
-        checkpointer_scheduler: &CheckpointScheduler<LR>,
+impl<C: LearnerComponents> LearnerCheckpointer<C> {
```
Comment (louisfd) [inline burn-train/src/learner/base.rs:33]: I'm starting to think there is too many times the letter C: callback, checkpoint, components

File: burn-train/src/learner/builder.rs [hunk 2]
```diff
@@ -98,16 +98,16 @@ where
     /// * `renderer` - The custom renderer.
     pub fn renderer<DR>(mut self, renderer: DR) -> Self
     where
-        DR: DashboardRenderer + 'static,
+        DR: MetricsRenderer + 'static,
```
Comment (louisfd) [inline burn-train/src/learner/builder.rs:104]: if there is no more dashboard, should not use D letter

**Review by louisfd — CHANGES_REQUESTED** [review]
Very nice. My only comments are about generics names

File: burn-train/src/learner/base.rs (hunk 1 above)
Comment (nathanielsimard) [reply burn-train/src/learner/base.rs:33]: We could use LC ?

**Pull request merged**